### ISSUE TITLE
Add generic typehints in PHPDoc

### ISSUE THREAD

**Issue by GuySartorelli** [silverstripe/silverstripe-framework#11066]
We should add PHPDoc typehints for generics where appropriate, to improve DX and type safety.

## Potentially Handy Links
- [project board of a fork where we experimented with this](https://github.com/orgs/silverstripeltd/projects/9/views/1)
- [devsense - Generics in PHP using PHP DocComments](https://blog.devsense.com/2022/generics-in-php-using-phpdoc)
- [phpstan - Generics by Examples](https://phpstan.org/blog/generics-by-examples)

## Acceptance criteria
- Don’t use phpstan/psalm prefixes (i.e. use `@return` and `@param` directly)
- Don't use any conditional return types
- Add generic typehints for all list implementations
- Add generic typehints for injector `get()`, `create()`, and any other methods where it would be relevant for injector
  - NOTE: Not for `Injectable` - i.e. `DataObject::create()` should just return `static` which it already does.
- Add generic typehints for methods which return different objects based on a classname that's passed in
  - e.g. `GridFieldConfig::getComponentByType()`
  - e.g. `SapphireTest::objFromFixture()`
- spend a little bit of time to see if there are other quick wins we can get by introducing generics.

## Notes
- These can be useful for static analysis testing (e.g. via phpstan or psalm) to validate return types are what you expect them to be - we should probably spin off a separate card to implement that. There's some devs internally who have experience with that.
- Once this is merged, we should look at how the [API documentation](https://api.silverstripe.org/4/SilverStripe.html) is rendering generics and do a follow card if needed. (see https://github.com/silverstripe/api.silverstripe.org/issues/115)

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/11108
- https://github.com/silverstripe/silverstripe-admin/pull/1648
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1433
- https://github.com/silverstripe/silverstripe-assets/pull/584
- https://github.com/silverstripe/silverstripe-auditor/pull/67
- https://github.com/silverstripe/silverstripe-blog/pull/739
- https://github.com/silverstripe/silverstripe-campaign-admin/pull/291
- https://github.com/silverstripe/silverstripe-cms/pull/2914
- https://github.com/silverstripe/silverstripe-contentreview/pull/220
- https://github.com/silverstripe/silverstripe-errorpage/pull/95
- https://github.com/silverstripe/silverstripe-graphql/pull/567
- https://github.com/silverstripe/silverstripe-lumberjack/pull/137
- https://github.com/silverstripe/recipe-cms/pull/79
- https://github.com/silverstripe/silverstripe-mfa/pull/524
- https://github.com/silverstripe/silverstripe-subsites/pull/554
- https://github.com/silverstripe/silverstripe-versioned/pull/431
- https://github.com/silverstripe/silverstripe-versioned-admin/pull/319
- https://github.com/silverstripe/silverstripe-userforms/pull/1261
- https://github.com/silverstripe/silverstripe-session-manager/pull/177
- https://github.com/silverstripe/silverstripe-sharedraftcontent/pull/222
- https://github.com/silverstripe/silverstripe-sitewidecontent-report/pull/76
- https://github.com/silverstripe/silverstripe-ldap/pull/67
- https://github.com/silverstripe/silverstripe-realme/pull/128
- https://github.com/silverstripe/silverstripe-securityreport/pull/72
- https://github.com/silverstripe/silverstripe-spamprotection/pull/108
- https://github.com/silverstripe/silverstripe-staticpublishqueue/pull/182
- https://github.com/silverstripe/silverstripe-versionfeed/pull/94
- https://github.com/silverstripe/silverstripe-gridfieldqueuedexport/pull/94
- https://github.com/silverstripe/silverstripe-login-forms/pull/162
- https://github.com/silverstripe/silverstripe-textextraction/pull/86
- https://github.com/silverstripe/silverstripe-elemental/pull/1131
- https://github.com/silverstripe/cwp-agencyextensions/pull/110
- https://github.com/silverstripe/cwp-core/pull/130
- https://github.com/silverstripe/cwp/pull/336
- https://github.com/silverstripe/silverstripe-iframe/pull/77
- https://github.com/silverstripe/silverstripe-registry/pull/90
- https://github.com/silverstripe/silverstripe-taxonomy/pull/102
- https://github.com/silverstripe/silverstripe-externallinks/pull/117
- https://github.com/silverstripe/silverstripe-tagfield/pull/277
- https://github.com/silverstripe/silverstripe-reports/pull/172
- https://github.com/silverstripe/silverstripe-testsession/pull/87
- https://github.com/silverstripe/silverstripe-config/pull/107
- https://github.com/silverstripe/silverstripe-siteconfig/pull/152
- https://github.com/tractorcow-farm/silverstripe-fluent/pull/819
- https://github.com/symbiote/silverstripe-advancedworkflow/pull/511
- https://github.com/symbiote/silverstripe-queuedjobs/pull/415
- https://github.com/silverstripe/developer-docs/pull/438

**Comment by emteknetnz** [comment]
I reviewed up to and including recipe-cms

**Comment by emteknetnz** [comment]
I reviewed up to and including login-forms
